#############################################################
# Description: A script for automatic RRBS Data Preprocessing
# Note: data was generated by NuGEN, therefore contains NuGEN specific Steps.
# Species: zebrafish
#############################################################
# Files needs to be prepared:
# 1: Zebrafish Genome, download the fa.gz file from Ensembl, MUST put it in a folder called Genome, and unzip is with gunzip XXX.fa.gz command
# 2: trimRRBSdiversityAdaptCustomers.py script, from https://github.com/nugentechnologies/NuMetRRBS
# 3. nudup.py script, from https://github.com/tecangenomics/nudup/blob/master/nudup.py
# 4. strip_bismark_sam.sh, from https://github.com/nugentechnologies/NuMetRRBS/blob/master/strip_bismark_sam.sh

FileFolder <- "../190822_D00623_0193_AH3Y2YBCX3"


message("[===============================]")
message("[<<<<< RRBS Analysis Start >>>>>]")
message("--------------------------------")

message("\n!!! This script is written for Medical Genomic Group to conduct NuGEN RRBS data analysis, so there are some commands and scipts specifically designed for NuGEN data, please DO NOT use it in all RRBS, like generated from other companies.!!!\n")


message("Your RRBS Data Folder is: ", FileFolder)
message("Your current folder is: ",getwd(), " folders contains middle result will be generated here.")

message("\n[ Section 1:  BCL2FASTQ ]\n")
if (!file.exists("./BCL2FASTQ")) dir.create("./BCL2FASTQ")
message("Created BCL2FASTQ in current folder for holding fasta files.")

cmd <- paste('bcl2fastq  --runfolder-dir ', FileFolder,' -p 40 --output-dir ./BCL2FASTQ --no-lane-splitting  --use-bases-mask Y*,I6Y* --minimum-trimmed-read-length 0 --mask-short-adapter-reads 0 &> ./BCL2FASTQ/bcl2fastq_log.txt',sep="")
message("--------------------------------")
message('[',cmd,']')
message("--------------------------------")
message("Above command would return both fastaq file for each sample, and their INDEX.fq file, which will be used in future NuDUP step.")


message("\nRunning...(you can check log in ./BCL2FASTQ/bcl2fastq_log.txt)")
system(cmd)

if (!file.exists("./BCL2FASTQ/INDEX")) dir.create("./BCL2FASTQ/INDEX")
system("mv ./BCL2FASTQ/*_R2_001.fastq.gz ./BCL2FASTQ/INDEX")

message("\n[ Section 2:  General Trming ]\n")

if (!file.exists("./GENERAL_TRIMING")) dir.create("./GENERAL_TRIMING")
message("Created GENERAL_TRIMING in current folder for holding fasta files.")

cmd <- "parallel --tmpdir ./PARALLEL_TMP --plus 'trim_galore -a AGATCGGAAGAGC -o ./GENERAL_TRIMING {...}.fastq.gz' &> ./GENERAL_TRIMING/general_triming.log ::: ./BCL2FASTQ/*.fastq.gz"
message('[',cmd,']')
message("--------------------------------")
message("Above command would do general triming on AGATCGGAAGAGC adapter.")


message("\nRunning...(you can check each output file in ./GENERAL_TRIMING folder.)")
system(cmd)

message("\n[ Section 3: NuGEND Diversity Triming ]\n")

if (!file.exists("./NuGEN_DIVERSITY_TRIMING")) dir.create("./NuGEN_DIVERSITY_TRIMING")
message("Created NuGEN_Diversity_TRIMING in current folder for holding fasta files.")

cmd <- "parallel --tmpdir ./PARALLEL_TMP --plus 'python ./trimRRBSdiversityAdaptCustomers.py -1 {...}.fq.gz' &> ./NuGEN_DIVERSITY_TRIMING/nugen_diversity_trming.log ::: ./GENERAL_TRIMING/*.fq.gz"
message("--------------------------------")
message('[',cmd,']')
message("--------------------------------")
message("Above command would run NuGEN specific diversity triming steps.")

message("\nRunning...")
system(cmd)
system("mv GENERAL_TRIMING/*trimmed.fq_trimmed.fq.gz ./NuGEN_DIVERSITY_TRIMING")

message("\n[ Section 4: Prepare Bismark Genome ]\n")
if (!file.exists("./Genome")) stop("Can not find Genome folder, you must have prepared genome.fa file as reference in it, unzipped.")

cmd <- "bismark_genome_preparation ./Genome &> ./Genome/bismark_preparation.log"
message("--------------------------------")
message('[',cmd,']')
message("--------------------------------")
message("Above command would be used to prepare genome for alignment.")

message("\nRunning...(you can check log in ./Genome/bismark_preparation.log)")
system(cmd)

message("\n[ Section 5: Bismark Alignment ]\n")
if (!file.exists("./BISMARK")) dir.create("./BISMARK")
message("Created BISMARK in current folder for holding fasta files.")

Samples <- dir('./NuGEN_DIVERSITY_TRIMING')
Samples <- Samples[sapply(Samples,function(x) "trimmed.fq.gz" %in% strsplit(x,split='_')[[1]])]

for(i in Samples) {
    message('\n-----------------------------------------------')
    message('Runing ',i,' with Bismark alignment now...')

    cmd <- paste('bismark --parallel 6 --output_dir ./BISMARK --bowtie2 ./Genome ./NuGEN_DIVERSITY_TRIMING/',i," &> ",i,'.log',sep='')
    #message(cmd)
    system(cmd,ignore.stdout=TRUE)
}
system("mkdir BISMARK/log")
system("mv ./*.log ./BISMARK/log/")

message("\n[ Section 6: Samtools Transfer ]\n")
if (!file.exists("./SAMTOOLS")) dir.create("./SAMTOOLS")
message("Created SAMTOOLS in current folder for holding sam files.")

cmd <- "parallel --tmpdir ./PARALLEL_TMP --plus 'samtools view -h -o ./SAMTOOLS/{/...}.sam {...}.fq_trimmed_bismark_bt2.bam' ::: ./BISMARK/*.bam"
message("--------------------------------")
message('[',cmd,']')
message("--------------------------------")
message("Above command would transfer BAM file generated by bismark to SAM files.")

message("\nRunning...")
system(cmd)


message("\n[ Section 7: Strip Bismark Result ]\n")

cmd <- "parallel --tmpdir ./PARALLEL_TMP --plus './strip_bismark_sam.sh {...}.sam' ::: ./SAMTOOLS/*.sam"
message("--------------------------------")
message('[',cmd,']')
message("--------------------------------")
message("Above command would strip SAM files generated by Samtools, which is provided by NuGEN company.")

message("\nRunning...")
system(cmd)


message("\n[ Section 8: Run Nudup ]\n")
if (!file.exists("./NUDUP")) dir.create("./NUDUP")
if (!file.exists("./NUDUP/TMP")) dir.create("./NUDUP/TMP")
message("Created NUDUP in current folder for holding deduplicated sam files.")

Samples <- dir('./BCL2FASTQ/INDEX')
Samples <- sapply(Samples,function(x) paste(strsplit(x,split="_")[[1]][1:2],collapse="_"))
Samples <- paste(Samples,collapse=" ")

cmd <- paste("parallel --tmpdir ./PARALLEL_TMP --plus ",
             " 'python ./nudup.py -f ./BCL2FASTQ/INDEX/{}_R2_001.fastq.gz -o ./NUDUP/{} -T ./NUDUP/TMP ./SAMTOOLS/{}_R1_001_trimmed.sam_stripped.sam' ::: ", 
             Samples, sep="")
message("--------------------------------")
message('[',cmd,']')
message("--------------------------------")
message("Above command would run NUDUP.")

message("\nRunning...")
system(cmd)

